src/main/scala/com/codacy/analysis/cli/upload/ResultsUploader.scala

```diff
@@ -63,27 +67,48 @@ class ResultsUploader private (commitUuid: String, codacyClient: CodacyClient, b
     codacyClient.sendEndOfResults(commitUuid)
   }
 
-  private def uploadResultsBatch(tool: String, batchSize: Int, results: Set[Result]): Future[Either[String, Unit]] = {
-    if (results.size <= batchSize) {
-      codacyClient.sendRemoteResults(tool, commitUuid, results)
-    } else {
-      val batch: Future[Either[String, Unit]] =
-        codacyClient.sendRemoteResults(tool, commitUuid, results.take(batchSize))
-      val remainingBatches: Future[Either[String, Unit]] =
-        uploadResultsBatch(tool, batchSize, results.drop(batchSize))
-      sequenceUploads(batch, remainingBatches)
-    }
+  private def uploadResultsBatch(tool: String,
+                                 batchSize: Int,
+                                 results: Set[FileResults]): Future[Either[String, Unit]] = {
+    val (_, remainingFileResults, uploadResponses) =
+      results.foldLeft((0, Set.empty[FileResults], Seq.empty[Future[Either[String, Unit]]])) {
+        case ((accumulatedResultsSize, accumulatedFileResults, responses), fileResults)
+            if (accumulatedResultsSize + fileResults.results.size) >= batchSize =>
+          val response = codacyClient.sendRemoteResults(tool, commitUuid, accumulatedFileResults)
+          (fileResults.results.size, Set(fileResults), responses :+ response)
+
+        case ((accumulatedResultsSize, accumulatedFileResults, responses), fileResults) =>
+          (accumulatedResultsSize + fileResults.results.size, accumulatedFileResults + fileResults, responses)
+      }
+    val response = codacyClient.sendRemoteResults(tool, commitUuid, remainingFileResults)
+
+    sequenceUploads(uploadResponses :+ response)
   }
 
-  private def sequenceUploads(call1: Future[Either[String, Unit]], call2: Future[Either[String, Unit]]) = {
-    for {
-      res1 <- call1
-      res2 <- call2
-    } yield {
-      Seq(res1, res2).collect {
-        case Left(error) => error
-      }.reduceOption(_ + "\n" + _).map(Left.apply).getOrElse(Right(()))
+  private def groupResultsByFile(files: Set[Path], results: Set[Result]): Set[FileResults] = {
+    val resultsByFile: Map[Path, Set[Result]] = results.groupBy {
+      case i: Issue      => i.filename
+      case fe: FileError => fe.filename
     }
+
+    val filesWithResults: Set[FileResults] = resultsByFile.map {
+      case (filename, fileResults) => FileResults(filename, fileResults)
+    }(collection.breakOut)
+
+    val filesWithoutResults: Set[FileResults] = files.diff(resultsByFile.keySet).map(FileResults(_, Set.empty[Result]))
+
+    filesWithResults ++ filesWithoutResults
```

Comment:
Instead of doing a diff to get filesWithoutResults and then combining filesWithoutResults with filesWithResults, I would do map over all files and find if they have results in filesWithResults.